### ISSUE TITLE
Global options in command handler: ParseResult.ValueForOption<T>(string) does not return anything

### ISSUE THREAD

**Issue by jhgbrt** [dotnet/command-line-api#1064]
I'm trying to set up a command that needs access to a global option. This is done through InvocationContext.

When retrieving an option from a ParseResult through the ValueForOption<T>(Option<T>) overload this works as expected, but when using the ValueForOption<T>(string optionName) overload, it doesn't find the option/value.

cf. unit test below
```
       [Fact]
       public void Testcase()
       {
           var child1 = new Command("child1");
           var fooOption = new Option<string>("--foo");
           child1.AddGlobalOption(fooOption);
           var root = new RootCommand();
           root.AddCommand(child1);
           var child2 = new Command("child2");
           var barOption = new Option<string>("--bar");
           child2.AddOption(barOption);
           string foo = null, bar = null;
           child2.Handler = CommandHandler.Create((InvocationContext context) => 
           {
               // this works:
               //foo = context.ParseResult.ValueForOption(fooOption);
               //bar = context.ParseResult.ValueForOption(barOption);

               //this doesn't:
               foo = context.ParseResult.ValueForOption<string>("--foo"); 
               bar = context.ParseResult.ValueForOption<string>("--bar");
           });
           child1.AddCommand(child2);

           root.Invoke("child1 --foo=FOO child2 --bar=BAR");

           Assert.Equal("FOO", foo);
           Assert.Equal("BAR", bar);
       }
```

**Comment by jonsequitur** [comment]
It seems intuitive that that should work and it's bitten me once or twice as well.

The problem is that `ValueForOption(string optionName)` doesn't do a recursive search of the parse tree. If it did, it would have to return a collection of results, since you could have two quite different options on different subcommands, e.g.:

```csharp
new RootCommand
{
    new Command("subcommand-one")
    {
        new Option<int>("--same")
    }
    new Command("subcommand-two")
    {
        new Option<string>("--same")
    }
}
```

Given this, you wouldn't know which one you were getting back, which wouldn't be very useful. So the string-based overload only searches the current level of the hierarchy.

In your example, something like this should work:

```csharp
foo = context.ParseResult.CommandResult.ValueForOption("--foo");
```

But it's slightly more performant, strongly-typed, and probably cleaner to use the `ValueForOption(Option)` overload.

**Comment by jhgbrt** [comment]
That is quite confusing... IMO, either these two overloads should be named differently, or they should behave similarly... 

I could imagine that the situation you're describing (multiple candidate options) could be considered an error and thus throw a descriptive exception?

**Comment by jonsequitur** [comment]
It is confusing and I suspect we should remove the string overload. We could replace it with a new method that does a complete search from the root but finding multiple candidate options for a given string will be a possibility.

**Comment by jonsequitur** [comment]
Yes, something like this can work. (Nitpick: a call to `IIdentifierSymbol.HasAlias` rather than the equality check in the `Where` would be more correct.)

The dictionaries are there to more easily enforce a single result per symbol and improve lookup speed especially when traversing the various levels of the syntax tree, since a given symbol (e.g. a global option) can be added at several different levels. This has been part of a general move away from string comparisons as it's clearer and faster.

**Comment by a-game** [comment]
I'm having the opposite problem in a way.

I have a command tree that looks like this
```csharp
var opt = new Option(new [] {"--foo", "-f"});
new RootCommand
{
    new Command("child")
    {
        new Command("grand-child") 
        {
            opt
        }
    }
}
```

And I'm trying to access the options in a middleware like this:
```csharp
 var parser = new CommandLineBuilder(cmd)
      .UseDefaults()
      .UseMiddleware(async (context, next) =>
      {
          // this  works with both -f and --foo as args.
          var foo = context.ParseResult.ValueForOption<string>(opt.Aliases.First());

          // does not work: returns ""
          var foo1 = context.ParseResult.ValueForOption<string>(opt);

          // Also does not work: returns null
          var optionResult = context.ParseResult.CommandResult.FindResultFor(opt);

          await next(context);
      }).Build();
```

I can see in  `ParseResult.cs` that the `T ValueForOption<T>(string alias)` looks at the `CommandResult.Children` while  ` T ValueForOption<T>(Option option)` only looks at the `RootCommand`

But I found it very confusing.

**Comment by jonsequitur** [comment]
@a-game, this is why I think this is confusing. The `string`-based lookups are inherently ambiguous, as discussed above.

This does work for me, though: ` var optionResult = context.ParseResult.CommandResult.FindResultFor(opt);` What version are you using?

The reason that `T ValueForOption<T>(Option option)` looks at the root command result is that that's where the indexes for the entire parse result are kept.
